File: benchmarks/Sources/mcp-benchmarker/KeyResidue.swift
import Foundation
import CryptoKit
#if canImport(Security)
import Security
#endif

// KeyResidue.swift — zero-residual-key verification for scratch estate
// retirement (P1.1b, the operator's 2026-07-30 requirement: when a test estate is
// retired, its key material goes with it — no orphaned keys ever).
//
// WHAT CAN LEAVE KEY MATERIAL BEHIND
//   - The Swift product's DURABLE posture mints one Keychain generic-password
//     item per estate (service "com.codedaptive.mootx01", account
//     "estate-db-key." + SHA-256 hex of the standardized estate file path —
//     contract with KeychainKeyStore.estateAccount(for:) in
//     PersistenceKitSQLite and EstateKeyProvider.keychainService in
//     MootInstallerCore; replicated here the same way ScratchPosture.swift
//     replicates the transient-record plaintext rule).
//   - A `db.key` file INSIDE the estate's own directory. This is how the Rust
//     product keys every estate (PersistenceKit rust encryption.rs
//     INSTALL_KEY_FILE), and how a harness build of the Swift product keys a
//     database the matrix converted and must reopen. Under a scratch dir,
//     teardown of the dir removes it.
//
// The harness's run modes are designed to leave NOTHING: unencrypted writes
// the transient-record rule (no key exists), and encrypted keeps every key out of
// the Keychain — either a temporal key in the serve process's memory, or a key
// file that dies with the scratch directory. This verifier is
// the enforcement: every retirement probes for both residue kinds, PURGES any
// Keychain item found (the key is removed along with the estate), and reports
// loudly — a residue hit means the posture contract was violated upstream and
// the run's cleanliness claim would otherwise be false.

/// Keychain service every mootx01 surface uses for estate db keys.
/// MUST match `EstateKeyProvider.keychainService` (MootInstallerCore).
let mootKeychainService = "com.codedaptive.mootx01"

/// The per-estate Keychain account for an estate file path.
/// MUST match `KeychainKeyStore.estateAccount(for:)` (PersistenceKitSQLite):
/// standardized path → SHA-256 → hex, prefixed "estate-db-key.".
func estateDbKeyAccount(forEstatePath path: String) -> String {
    let standardized = URL(fileURLWithPath: path).standardizedFileURL.path
    let digest = SHA256.hash(data: Data(standardized.utf8))
    let hex = digest.map { String(format: "%02x", $0) }.joined()
    return "estate-db-key.\(hex)"
}

/// What a scratch dir could leave behind, collected BEFORE teardown (the
/// estate files must still exist to enumerate them).
struct KeyResidueProbes: Sendable, Equatable {
    /// Keychain accounts derived from every estate database file found under
    /// the scratch dir (Swift-product residue candidates).
    let keychainAccounts: [String]
    /// On-disk key files found under the scratch dir (Rust-product `db.key`
    /// convention). Recorded as paths so post-teardown absence can be asserted
    /// even if the whole-dir removal partially failed.
    let keyFilePaths: [String]
    /// The scratch dir itself.
    let scratchDirPath: String
    /// Login-keychain item count under the ESTATE IDENTITY service
    /// (`com.mootx01.estate.identity`) at probe time. Identity items are
    /// keyed by the estate's own UUID, which the harness never learns, so
    /// residue is detected by GROWTH across the estate's lifetime rather
    /// than by account. Growth means the product wrote a scratch estate's
    /// Ed25519 identity key into the real Keychain — the exact leak that
    /// accumulated 971 orphans by 2026-08-06 (fixed product-side the same
    /// day: ephemeral lifetime forces the in-memory identity store, and the
    /// unencrypted lane now declares ephemeral).
    let identityItemCountBefore: Int
}

/// Walks a scratch dir and records every residue candidate: one Keychain
/// account per estate database file (`*.sqlite` / `*.sqlite3`), plus every
/// on-disk `db.key`. Call BEFORE tearing the dir down. A missing dir yields
/// empty probes (nothing was created, nothing can linger).
///
/// - Precondition: `scratchDir` must be a harness-created temporary
///   directory. The Keychain purge in `verifyZeroKeyResidue` is
///   path-specific (accounts derive from SHA-256 of paths under this dir),
///   so probing a real estate's directory would target its real key.
func collectKeyResidueProbes(scratchDir: URL) -> KeyResidueProbes {
    var accounts: [String] = []
    var keyFiles: [String] = []
    let fm = FileManager.default
    if let walker = fm.enumerator(at: scratchDir, includingPropertiesForKeys: nil) {
        for case let url as URL in walker {
            let name = url.lastPathComponent
            if name.hasSuffix(".sqlite") || name.hasSuffix(".sqlite3") {
                accounts.append(estateDbKeyAccount(forEstatePath: url.path))
            }
            if name == "db.key" {
                keyFiles.append(url.path)
            }
        }
    }
    // The default estate path is probed even when the walk found no estate
    // file (e.g. a run that died before first write): a Keychain item keyed to
    // the path that WOULD have been created is still residue worth checking.
    let defaultEstate = scratchDir.appendingPathComponent("estate.sqlite").path
    let defaultAccount = estateDbKeyAccount(forEstatePath: defaultEstate)
    if !accounts.contains(defaultAccount) {
        accounts.append(defaultAccount)
    }
    return KeyResidueProbes(
        keychainAccounts: accounts.sorted(),
        keyFilePaths: keyFiles.sorted(),
        scratchDirPath: scratchDir.path,
        identityItemCountBefore: mootIdentityItemCount())
}

/// The outcome of a zero-residual verification.
struct KeyResidueReport: Sendable, Equatable {
    /// Keychain accounts that still held an item after retirement. Every entry
    /// here was PURGED by the verifier (the key is removed along with the
    /// estate); a non-empty list still marks the run as having violated the
    /// zero-residue contract upstream.
    var keychainItemsFoundAndPurged: [String] = []
    /// Keychain accounts whose item could not be removed (purge failed).
    var keychainItemsUnremovable: [String] = []
    /// On-disk key files still present after teardown.
    var keyFilesRemaining: [String] = []
    /// True when the scratch dir itself survived teardown.
    var scratchDirRemaining = false
    /// How many estate-identity Keychain items appeared during this estate's
    /// lifetime (growth over the probe-time count). Non-purgeable — identity
    /// items are keyed by the estate's own UUID, which the harness never
    /// learns — so a non-zero value is a loud product-fix regression signal,
    /// not a cleanup action.
    var identityItemsGrownBy = 0
    /// Estate database files that were expected plaintext but carry an
    /// encrypted header — the at-rest posture silently flipped (a product
    /// binary that predates the ephemeral+marker fix): the cell would be
    /// MISLABELED as unencrypted.
    var postureViolations: [String] = []

    /// Zero residual key material: nothing found, nothing left behind.
    var isClean: Bool {
        keychainItemsFoundAndPurged.isEmpty
            && keychainItemsUnremovable.isEmpty
            && keyFilesRemaining.isEmpty
            && !scratchDirRemaining
            && identityItemsGrownBy == 0
            && postureViolations.isEmpty
    }
}

/// Keychain service every mootx01 surface uses for estate Ed25519 identity
/// keys. MUST match `EstateIdentityKeyStore.service` in LocusKit.
let mootIdentityKeychainService = "com.mootx01.estate.identity"

/// Counts login-keychain items under the estate-identity service. 0 when
/// Security is unavailable or the query fails — the growth check then
/// degrades to a no-op rather than a false alarm.
func mootIdentityItemCount() -> Int {
    #if canImport(Security)
    let query: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: mootIdentityKeychainService,
        kSecMatchLimit as String: kSecMatchLimitAll,
        kSecReturnAttributes as String: true,
    ]
    var result: CFTypeRef?
    guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
          let items = result as? [[String: Any]] else { return 0 }
    return items.count
    #else
    return 0
    #endif
}

#if canImport(Security)
/// True when a generic-password item exists for the moot service + account.
func keychainItemExists(account: String) -> Bool {
    let query: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: mootKeychainService,
        kSecAttrAccount as String: account,
        kSecReturnAttributes as String: false,
    ]
    return SecItemCopyMatching(query as CFDictionary, nil) == errSecSuccess
}

/// Deletes the generic-password item for the moot service + account.
/// Returns true when the item is gone afterwards (deleted, or never existed).
@discardableResult
func deleteKeychainItem(account: String) -> Bool {
    let query: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: mootKeychainService,
        kSecAttrAccount as String: account,
    ]
    let status = SecItemDelete(query as CFDictionary)
    return status == errSecSuccess || status == errSecItemNotFound
}
#endif

/// Verifies zero residual key material AFTER a scratch dir teardown, purging
/// any Keychain item it finds (requirement: the key is removed along with the
/// estate). Pure inspection plus purge — never throws; the caller decides how
/// loud to be from the report.
func verifyZeroKeyResidue(after probes: KeyResidueProbes) -> KeyResidueReport {
    var report = KeyResidueReport()
    let fm = FileManager.default
    report.scratchDirRemaining = fm.fileExists(atPath: probes.scratchDirPath)
    for keyFile in probes.keyFilePaths where fm.fileExists(atPath: keyFile) {
        report.keyFilesRemaining.append(keyFile)
    }
    #if canImport(Security)
    for account in probes.keychainAccounts where keychainItemExists(account: account) {
        if deleteKeychainItem(account: account), !keychainItemExists(account: account) {
            report.keychainItemsFoundAndPurged.append(account)
        } else {
            report.keychainItemsUnremovable.append(account)
        }
    }
    // Identity-item growth check: see KeyResidueProbes.identityItemCountBefore.
    report.identityItemsGrownBy =
        max(0, mootIdentityItemCount() - probes.identityItemCountBefore)
    #endif
    return report
}

/// Reports a scratch estate deliberately KEPT because its scope did not finish
/// cleanly, naming the path so the work can be inspected, salvaged, or removed
/// by hand.
///
/// Retirement is for a scope that SUCCEEDED. A scope that threw leaves whatever
/// it built where it lies. On 2026-08-17 an unconditional teardown deleted a
/// finished 100,000-row landscape because the call that would have returned it
/// was slow: 100 minutes of built and encoded work, destroyed by the error path
/// of the thing that had already done the work. Disk is cheap and a rebuild is
/// not, so an error keeps its evidence and the operator decides what happens
/// to it.
///
/// For an encrypted scratch estate this leaves key material on disk, which is
/// the deliberate trade: the run has already failed, and a kept estate under a
/// loud notice beats silently destroying hours of work.
func keepScratchEstateOnFailure(_ scratchDir: URL, lane: String) {
    // Mark the kept scratch with a `.kept` suffix so it is distinguishable
    // from ACTIVE scratches of concurrently running lanes (which share the
    // /tmp prefix); the pruning below must never touch a live scratch.
    let fm = FileManager.default
    var kept = scratchDir
    let marked = URL(fileURLWithPath: scratchDir.path + ".kept")
    if (try? fm.moveItem(at: scratchDir, to: marked)) != nil { kept = marked }
    FileHandle.standardError.write(Data(
        ("[\(lane)] scope did not finish cleanly — scratch estate KEPT at \(kept.path)\n"
         + "[\(lane)] it holds everything built up to the failure; remove it by hand "
         + "when you are done with it\n").utf8))

    // Cap the kept population. The keep-the-evidence rationale above
    // protects RECENT failures; a hundred identical transient-failure
    // scratches carry no additional evidence, and 181 accumulated keeps
    // produced a 61 GB disk crisis on 2026-08-26. Keep the newest 5 `.kept`
    // dirs sharing this scratch's family prefix (path up to the random
    // suffix); announce every prune so nothing disappears silently.
    let familyPrefix = kept.deletingLastPathComponent().path + "/"
        + kept.lastPathComponent.split(separator: "-").dropLast().joined(separator: "-") + "-"
    let parent = kept.deletingLastPathComponent()
    guard let siblings = try? fm.contentsOfDirectory(
        at: parent, includingPropertiesForKeys: [.contentModificationDateKey]) else { return }
    let keptFamily = siblings.filter {
        $0.path.hasPrefix(familyPrefix) && $0.path.hasSuffix(".kept")
    }.sorted { a, b in
        let da = (try? a.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
        let db = (try? b.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
        return da > db
    }
    for old in keptFamily.dropFirst(5) {
        try? fm.removeItem(at: old)
        FileHandle.standardError.write(Data(
            ("[\(lane)] pruned old kept scratch (cap 5): \(old.path)\n").utf8))
    }
}

/// Retires one scratch estate: collect residue probes, run the guarded
/// teardown the caller supplies, then verify + purge. A dirty report is
/// shouted to stderr with every finding named — a residue hit means the
/// posture contract was violated upstream, and silence would let a run claim
/// cleanliness it does not have.
///
/// - Precondition: `scratchDir` must be a harness-created temporary
///   directory (see `collectKeyResidueProbes`).
func retireScratchEstate(
    _ scratchDir: URL,
    expectPlaintext: Bool = false,
    teardown: (URL) throws -> Void
) rethrows {
    let probes = collectKeyResidueProbes(scratchDir: scratchDir)

    // Posture check runs PRE-teardown (the files must still exist). A
    // plaintext SQLite file begins with the 16-byte magic "SQLite format 3\0";
    // an SQLCipher-encrypted file does not. `expectPlaintext` comes from the
    // lane's declared posture — a mismatch means the product silently flipped
    // the at-rest posture (e.g. a binary that predates the ephemeral+marker
    // fix) and the cell would be mislabeled as unencrypted.
    var postureViolations: [String] = []
    if expectPlaintext {
        let magic = Data("SQLite format 3".utf8) + Data([0])
        let fm = FileManager.default
        if let walker = fm.enumerator(at: scratchDir, includingPropertiesForKeys: nil) {
            for case let url as URL in walker
            where url.lastPathComponent.hasSuffix(".sqlite")
                || url.lastPathComponent.hasSuffix(".sqlite3") {
                if let handle = try? FileHandle(forReadingFrom: url),
                   let head = try? handle.read(upToCount: magic.count),
                   head != magic {
                    postureViolations.append(url.path)
                }
            }
        }
    }

    // Instrument seam (W2.5 Track R(c)): copy the working estate out for
    // the optimizer's trace aggregation BEFORE teardown destroys it. No-op
    // unless MOOT_BENCH_KEEP_ESTATES_DIR is set; retirement's residue
    // verification below still runs on the scratch itself.
    keepEstateSeam(from: scratchDir)

    try teardown(scratchDir)
    var report = verifyZeroKeyResidue(after: probes)
    report.postureViolations = postureViolations
    guard !report.isClean else { return }
    var lines = ["[key-residue] RESIDUAL KEY MATERIAL at retirement of \(scratchDir.path):"]
    for a in report.keychainItemsFoundAndPurged {
        lines.append("  keychain item FOUND (now purged): service=\(mootKeychainService) account=\(a)")
    }
    for a in report.keychainItemsUnremovable {
        lines.append("  keychain item FOUND and NOT REMOVABLE: service=\(mootKeychainService) account=\(a)")
    }
    for f in report.keyFilesRemaining {
        lines.append("  key file still on disk: \(f)")
    }
    if report.scratchDirRemaining {
        lines.append("  scratch dir still present: \(probes.scratchDirPath)")
    }
    if report.identityItemsGrownBy > 0 {
        lines.append("  estate-identity keychain items GREW by \(report.identityItemsGrownBy) "
            + "(service=\(mootIdentityKeychainService)) — the product wrote a scratch estate's "
            + "identity key to the real Keychain; requires a product binary at or after the "
            + "2026-08-06 ephemeral-identity fix. NOT purgeable by account from here.")
    }
    for p in report.postureViolations {
        lines.append("  POSTURE FLIP: expected plaintext but encrypted header at \(p) "
            + "— the cell is mislabeled; a transient record (--db <scratch>) must be "
            + "created plaintext by the product.")
    }
    lines.append("  zero-residue contract violated — investigate the run's estate mode plumbing.")
    FileHandle.standardError.write(Data((lines.joined(separator: "\n") + "\n").utf8))
}

File: benchmarks/rust/src/key_residue.rs
// key_residue.rs — zero-residual-key verification for scratch estate
// retirement (P1.1b, the 2026-07-30 requirement: when a test estate is
// retired, its key material goes with it — no orphaned keys ever).
//
// The Rust leg targets Linux/Windows as well as macOS, so it verifies the
// ON-DISK residue kind only: the Rust product keys its estates from a
// `db.key` file INSIDE the estate's own directory (PersistenceKit rust
// encryption.rs, INSTALL_KEY_FILE) — under a scratch dir, teardown of the
// dir removes it, and this module proves that it did. The Swift product's
// Keychain residue kind is verified by the Swift leg (KeyResidue.swift);
// there is no Keychain on the Rust leg's target platforms.
//
// The harness's run modes are designed to leave NOTHING: unencrypted opens a
// plaintext transient record (no key exists), encrypted uses the temporal-key
// posture (Swift product: key lives only in the serve process's memory;
// Rust product: scratch-local db.key torn down with the dir). This verifier
// is the enforcement — every retirement probes for residue and reports
// loudly, because a residue hit means the posture contract was violated
// upstream.

use std::path::{Path, PathBuf};

use crate::mcp_client::MCPError;

/// What a scratch dir could leave behind, collected BEFORE teardown (the
/// files must still exist to enumerate them).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyResidueProbes {
    /// On-disk `db.key` files found under the scratch dir.
    pub key_file_paths: Vec<PathBuf>,
    /// The scratch dir itself.
    pub scratch_dir: PathBuf,
}

/// Walks a scratch dir and records every on-disk key file (`db.key`). Call
/// BEFORE tearing the dir down. A missing dir yields empty probes (nothing
/// was created, nothing can linger).
pub fn collect_key_residue_probes(scratch_dir: &Path) -> KeyResidueProbes {
    let mut key_files = Vec::new();
    collect_db_keys(scratch_dir, &mut key_files, 0);
    key_files.sort();
    KeyResidueProbes {
        key_file_paths: key_files,
        scratch_dir: scratch_dir.to_path_buf(),
    }
}

/// Maximum depth this walk will descend below the scratch dir.
///
/// A scratch estate is shallow — `<scratch>/db.key` sits at depth 0 below
/// the scratch root — so this bound is never reached in practice. It is
/// defence in depth: teardown runs on every retirement and must not become an
/// unbounded walk over a pathological tree. Hitting the bound stops the
/// descent; it is not an error, because a residue walk that refuses to finish
/// is worse than one that stops looking.
const MAX_RESIDUE_WALK_DEPTH: usize = 16;

fn collect_db_keys(dir: &Path, out: &mut Vec<PathBuf>, depth: usize) {
    if depth >= MAX_RESIDUE_WALK_DEPTH {
        return;
    }
    let Ok(entries) = std::fs::read_dir(dir) else {
        return;
    };
    for entry in entries.flatten() {
        // `DirEntry::file_type()` describes the ENTRY (an lstat), so a symlink
        // reports `is_symlink()` and never `is_dir()`. `Path::is_dir()` stats
        // the TARGET and therefore follows symlinks — using it here let a
        // symlinked directory inside the scratch dir send this walk outside
        // the scratch estate. That matters because scratch paths are
        // deterministic under a shared /tmp and are accepted via
        // `create_dir_all()` when already present, so a same-user process can
        // pre-create one containing a symlink to a large tree or to `/`, and
        // every teardown would then walk it.
        //
        // A symlink is still eligible to be RECORDED as residue below when it
        // is named `db.key` — a link left in the scratch dir is residue — it
        // is only never DESCENDED into.
        let Ok(file_type) = entry.file_type() else {
            continue;
        };
        let path = entry.path();
        if file_type.is_dir() {
            collect_db_keys(&path, out, depth + 1);
        } else if path.file_name().and_then(|n| n.to_str()) == Some("db.key") {
            out.push(path);
        }
    }
}

/// The outcome of a zero-residual verification.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct KeyResidueReport {
    /// On-disk key files still present after teardown.
    pub key_files_remaining: Vec<PathBuf>,
    /// True when the scratch dir itself survived teardown.
    pub scratch_dir_remaining: bool,
}

impl KeyResidueReport {
    /// Zero residual key material: nothing left behind.
    pub fn is_clean(&self) -> bool {
        self.key_files_remaining.is_empty() && !self.scratch_dir_remaining
    }
}

/// Verifies zero residual key material AFTER a scratch dir teardown. Pure
/// inspection — the caller decides how loud to be from the report.
pub fn verify_zero_key_residue(probes: &KeyResidueProbes) -> KeyResidueReport {
    KeyResidueReport {
        key_files_remaining: probes
            .key_file_paths
            .iter()
            .filter(|p| p.exists())
            .cloned()
            .collect(),
        scratch_dir_remaining: probes.scratch_dir.exists(),
    }
}

/// Retires one scratch estate: collect residue probes, run the guarded
/// teardown the caller supplies, then verify. A dirty report is shouted to
/// stderr with every finding named — silence would let a run claim
/// cleanliness it does not have. Twin of Swift `retireScratchEstate`.
pub fn retire_scratch_estate<F>(scratch_dir: &Path, teardown: F) -> Result<(), MCPError>
where
    F: FnOnce(&Path) -> Result<(), MCPError>,
{
    let probes = collect_key_residue_probes(scratch_dir);
    let teardown_result = teardown(scratch_dir);
    let report = verify_zero_key_residue(&probes);
    if !report.is_clean() {
        let mut lines = vec![format!(
            "[key-residue] RESIDUAL KEY MATERIAL at retirement of {}:",
            scratch_dir.display()
        )];
        for f in &report.key_files_remaining {
            lines.push(format!("  key file still on disk: {}", f.display()));
        }
        if report.scratch_dir_remaining {
            lines.push(format!(
                "  scratch dir still present: {}",
                probes.scratch_dir.display()
            ));
        }
        lines.push(
            "  zero-residue contract violated — investigate the run's estate mode plumbing."
                .to_string(),
        );
        eprintln!("{}", lines.join("\n"));
    }
    teardown_result
}

#[cfg(test)]
mod tests {
    use super::*;

    fn scratch(name: &str) -> PathBuf {
        let dir = std::env::temp_dir().join(format!("key-residue-test-{name}-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(dir.join("estates").join("default")).unwrap();
        dir
    }

    #[test]
    fn probes_find_nested_db_key() {
        let dir = scratch("probe");
        std::fs::write(dir.join("estates/default/db.key"), b"0123456789abcdef0123456789abcdef")
            .unwrap();
        let probes = collect_key_residue_probes(&dir);
        assert_eq!(probes.key_file_paths.len(), 1);
        assert!(probes.key_file_paths[0].ends_with("estates/default/db.key"));
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn retirement_removes_key_with_estate_and_reports_clean() {
        let dir = scratch("clean");
        std::fs::write(dir.join("estates/default/db.key"), b"0123456789abcdef0123456789abcdef")
            .unwrap();
        std::fs::write(dir.join("estates/default/estate.sqlite"), b"data").unwrap();
        let probes = collect_key_residue_probes(&dir);
        assert_eq!(probes.key_file_paths.len(), 1);
        // Teardown = whole-dir removal, exactly what the runners do.
        std::fs::remove_dir_all(&dir).unwrap();
        let report = verify_zero_key_residue(&probes);
        assert!(report.is_clean(), "expected zero residue, got {report:?}");
    }

    #[test]
    fn residual_key_after_failed_teardown_is_detected() {
        let dir = scratch("dirty");
        std::fs::write(dir.join("estates/default/db.key"), b"0123456789abcdef0123456789abcdef")
            .unwrap();
        let probes = collect_key_residue_probes(&dir);
        // Simulate a teardown that failed to remove anything.
        let report = verify_zero_key_residue(&probes);
        assert!(!report.is_clean());
        assert_eq!(report.key_files_remaining.len(), 1);
        assert!(report.scratch_dir_remaining);
        let _ = std::fs::remove_dir_all(&dir);
    }

    /// REGRESSION (MXE-BK defect 3). A symlinked directory inside the scratch
    /// dir must NOT be descended into. `Path::is_dir()` stats the target and
    /// so followed the link, taking every teardown walk outside the scratch
    /// estate; `DirEntry::file_type()` describes the entry and does not.
    ///
    /// The outside tree here holds a `db.key`. Pre-fix the walk followed the
    /// link and collected it; post-fix it is never reached — the walk finds
    /// only the estate's own key.
    #[test]
    fn walk_does_not_follow_symlinked_directory_out_of_scratch() {
        let dir = scratch("symlink");
        std::fs::write(dir.join("estates/default/db.key"), b"inside").unwrap();

        // A tree OUTSIDE the scratch dir, holding a key the walk must not see.
        let outside = std::env::temp_dir()
            .join(format!("key-residue-outside-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&outside);
        std::fs::create_dir_all(&outside).unwrap();
        std::fs::write(outside.join("db.key"), b"outside").unwrap();

        #[cfg(unix)]
        std::os::unix::fs::symlink(&outside, dir.join("escape")).unwrap();
        #[cfg(windows)]
        std::os::windows::fs::symlink_dir(&outside, dir.join("escape")).unwrap();

        let probes = collect_key_residue_probes(&dir);

        assert_eq!(
            probes.key_file_paths.len(),
            1,
            "walk escaped the scratch estate through a symlink: {:?}",
            probes.key_file_paths
        );
        assert!(probes.key_file_paths[0].ends_with("estates/default/db.key"));
        assert!(
            !probes.key_file_paths.iter().any(|p| p.starts_with(&outside)),
            "walk collected a key from outside the scratch dir"
        );

        let _ = std::fs::remove_dir_all(&dir);
        let _ = std::fs::remove_dir_all(&outside);
    }

    /// A symlink named `db.key` is still RECORDED as residue — it is a link
    /// left inside the scratch dir. Only descent is refused, not detection.
    #[test]
    fn symlinked_key_file_is_still_recorded_as_residue() {
        let dir = scratch("symlink-key");
        let target = dir.join("real.key");
        std::fs::write(&target, b"k").unwrap();

        #[cfg(unix)]
        std::os::unix::fs::symlink(&target, dir.join("estates/default/db.key")).unwrap();
        #[cfg(windows)]
        std::os::windows::fs::symlink_file(&target, dir.join("estates/default/db.key")).unwrap();

        let probes = collect_key_residue_probes(&dir);
        assert_eq!(probes.key_file_paths.len(), 1);
        assert!(probes.key_file_paths[0].ends_with("estates/default/db.key"));

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn missing_dir_probes_empty_and_clean() {
        let dir = std::env::temp_dir().join("key-residue-test-never-created");
        let probes = collect_key_residue_probes(&dir);
        assert!(probes.key_file_paths.is_empty());
        assert!(verify_zero_key_residue(&probes).is_clean());
    }

    #[test]
    fn retire_runs_teardown_and_propagates_result() {
        let dir = scratch("retire");
        std::fs::write(dir.join("estates/default/db.key"), b"k").unwrap();
        let result = retire_scratch_estate(&dir, |d| {
            std::fs::remove_dir_all(d).map_err(|e| MCPError {
                description: format!("teardown: {e}"),
            })
        });
        assert!(result.is_ok());
        assert!(!dir.exists());
    }
}
